pkg/storage/async_store.go

```diff
@@ -49,8 +52,25 @@ func (a *AsyncStore) GetChunkRefs(ctx context.Context, userID string, from, thro
 	var ingesterChunks []string
 
 	go func() {
+		ingesterQueryFrom := from
+		if a.queryIngestersWithin != 0 {
+			oldestIngesterQueryFrom := model.Now().Add(-a.queryIngestersWithin)
+			// don't query ingesters if the query end is not after oldest allowed start time.
+			if !through.After(oldestIngesterQueryFrom) {
+				level.Debug(pkg_util.Logger).Log("msg", "skipping querying ingesters for chunk ids", "query-from", from, "query-through", through)
+				errs <- nil
+				return
+			}
+
+			// if ingesterQueryFrom is older than oldest allowed ingester query start then reduce it to oldest allowed start time.
+			if ingesterQueryFrom.Before(oldestIngesterQueryFrom) {
```

Comment:
I don't think we actually need this case. I don't think we should be altering the queries unnecessarily here as it only adds future complexity.

There is the odd case where `from` lies outside of the `query-ingesters-within` range, but the `to` is inside the range. In that case, I think we should be able to optimistically return `from---->to`, not just `query-ingesters-within----->to`.

Reply:
Wouldn't that mean honouring `query-ingesters-within` only when both `from` and `to` are outside of that range? 
Most of the queries we would get would overlap with `query-ingesters-within` so I don't think we would be saving much by not reducing the query interval for ingesters.
What do you think?

Reply:
I think the end result is equivalent for deployments with the query-frontend enabled and irrelevant for ones without the query-frontend (which don't use sharding code). The main goal of `query-ingesters-within` is not to prevent querying for data, but to prevent DoS'ing the ingesters. The flag is intended to be set to a large enough value that the ingesters wouldn't  hold any data in that range anyways (it would already be flushed). The reason I prefer to remove this clause is because it seems unnecessary as Loki currently stands and adds complexity that may hurt us in the future.

Reply:
Well, we already limit the query interval considering `query-ingesters-within` for in-memory logs of ingesters.
We will have to change that too to always query the whole query duration when it overlaps with `query-ingesters-within` duration. See https://github.com/grafana/loki/blob/61ba02e2c477896075f61d795e3a4fb90e87201a/pkg/querier/querier.go#L164

Reply:
>  we already limit the query interval considering query-ingesters-within for in-memory logs of ingesters.

That's actually my point, too. We already adjust the query intervals there, so there's no need to add extra complexity here in the async store -- the start/end times will _already_ be adjusted. I think you can remove this current code _safely_ and I think we benefit from the decreased complexity here.

Reply:
The async store actually executes the store query which actually gets the query for the whole query range.
In microservices mode, only the ingester query range is altered by limiting it to `query-ingesters-within` duration while the store query covers the whole query range.
I understand it is complex. We do all those to tune things to ease things a bit for Loki users. I have added comments wherever needed and can add more if you want.

Reply:
Hm. I don't think this actually makes a difference either way (removing this case or not), so I'm curious if there's a scenario where it's _helpful_ to have the extra complexity here. Can you give an example?

Reply:
I had deployed this in Loki ops cluster and noticed a drop of 1000 QPS which also somewhat helped reduce the ingester CPU usage as well. See https://admin-ops-tools1.grafana.net/grafana/goto/Pek3X3BMz

Reply:
That makes sense, but I expect you'd see the _same benefit_ if you removed this clause as well.

Reply:
yep, you are right. We are getting the same benefit even after removing this change.
I have reverted it and it does make sense to not have it. Sorry for all the back and forth.

Reply:
No worries, thanks for putting up with me :)